File: assignment_3/src/parser.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Parser {
  private static Scanner scanner;
  private static String token;

  private static Boolean factor() {
    token = scanner.next();
    System.out.println("factor " + token);
    return token.equals("id") || token.equals("num");
  }

  private static Boolean expression_prime() {
    token = scanner.next();
    System.out.println("expression_prime " + token);
    if (token.equals("+") || token.equals("-")) {
      return factor();
    }
    return true;
  }

  private static Boolean expression() {
    if (factor()) {
      return expression_prime();
    }
    return false;
  }

  private static Boolean statement() {
    token = scanner.next();
    System.out.println("statement_1 " + token);
    if (!token.equals("id")) {
      return false;
    }
    token = scanner.next();
    System.out.println("statement_2 " + token);
    if (!token.equals("=")) {
      return false;
    }
    return expression();
  }

  private static Boolean statement_list_prime() {
    if (!statement_list()) {
      return token.equals("$");
    }
    return false;
  }

  private static Boolean statement_list() {
    if (statement()) {
      if (!token.equals(";")) {
        return false;
      }
      return statement_list_prime();
    }
    return false;
  }

  private static Boolean program() {
    token = scanner.next();
    System.out.println("program " + token);
    if (token.equals("begin") && statement_list()) {
      token = scanner.next();
      System.out.println("program " + token);
      return token.equals("end");
    }
    return false;
  }

  public static void main(String[] args) {
    if (args.length < 1) {
      System.out.println("Missing input file argument");
      return;
    }
    try {
      scanner = new Scanner(new File(args[0]));

    } catch (FileNotFoundException e) {
      System.out.println("Could not locate file: " + args[0]);
      return;
    }

    //try {
      if (program()) {
        System.out.println("SUCCESS");
      } else {
        System.out.println("ERROR");
      }
      scanner.close();
    //} catch (Exception e) {
    //  System.out.println("ERROR");
    //}
  }
}

File: assignment_3/src/parser.py
file = open('../data/input1.txt', 'r')
global token

def get_token():
    token = file.readline().strip()
    print token
    return token

def factor():
    print "factor"
    return get_token() in ["id", "num"]

def expression_prime():
    print "expression_prime"
    token = get_token()
    if token in ["+", "-"] and factor():
        token = get_token()
        return token == ";"
    return token == ";"

def expression():
    print "expression"
    if not factor():
        return False
    if not expression_prime():
        return False
    return True

def statement():
    print "statement"
    if get_token() != "id":
        return False
    if get_token() != "=":
        return False
    if not expression():
        return False
    return True

def statement_list_prime():
    print "statement_list_prime"
    if not statement_list():
        return token == "end"
    return True

def statement_list():
    print "statement_list 1"
    if not statement():
        return False
    if not statement_list_prime():
        return False
    return True

def program():
    if get_token() != "begin":
        return False
    if not statement_list():
        return False
    if get_token() != "end":
        return False
    return True

if __name__ == '__main__':
    print program()
